backend/ftp/dataconn_test.go

```diff
@@ -0,0 +1,158 @@
+
+	// dataconn is nil - open for read
+	ftpfile.dataconn = nil
+	client.EXPECT().
+		RetrFrom(ftpfile.Path(), uint64(0)).
+		Return(&ftp.Response{}, nil).
+		Once()
+	dc, err = getDataConn(context.Background(), ftpfile, types.OpenRead)
+	s.NoError(err, "no error expected")
+	s.IsTypef(&dataConn{}, dc, "dataconn returned")
+
+	// dataconn is nil - error getting client
+	ftpfile.dataconn = nil
+	defaultClientGetter = clientGetterReturnsError
+	ftpfile.fileSystem.ftpclient = nil
+	dc, err = getDataConn(context.Background(), ftpfile, types.OpenRead)
+	s.Error(err, "error is expected")
+	s.ErrorIs(err, errClientGetter, "error is right kind of error")
+	s.Nil(dc, "dataconn should be nil on error")
+	defaultClientGetter = getClient
+	ftpfile.fileSystem.ftpclient = client
+
+	// dataconn is nil - error calling client.RetrFrom
+	someErr := errors.New("some error")
+	client.EXPECT().
+		RetrFrom(ftpfile.Path(), uint64(0)).
+		Return(nil, someErr).
+		Once()
+	dc, err = getDataConn(context.Background(), ftpfile, types.OpenRead)
+	s.Error(err, "error is expected")
+	s.ErrorIs(err, someErr, "error is right kind of error")
+	s.Nil(dc, "dataconn should be nil on error")
+
+	// dataconn is nil - open for write - error calling client.StorFrom
+	ftpfile.dataconn = nil
+	client.EXPECT().
+		StorFrom(ftpfile.Path(), mock.Anything, uint64(0)).
+		Return(someErr).
+		Once()
+	dc, err = getDataConn(context.Background(), ftpfile, types.OpenWrite)
+	s.NoError(err, "no error expected")
+	// error in getDataConn should close the PipeReader meaning Write errors
+	_, err = dc.Write([]byte{})
+	s.Error(err, "error is expected")
+
+	// dataconn is nil - open for write - success
+	ftpfile.dataconn = nil
+	client.EXPECT().
+		StorFrom(ftpfile.Path(), mock.Anything, uint64(0)).
+		Return(nil).
+		Once()
+	dc, err = getDataConn(context.Background(), ftpfile, types.OpenWrite)
+	s.NoError(err, "no error expected")
+	s.IsTypef(&dataConn{}, dc, "dataconn returned")
+
+	// open dataconn for read after dataconn for write exists - error on dataconn.Close
+	fakedconn := NewFakeDataConn(types.OpenWrite)
+	closeErr := errors.New("some close err")
+	fakedconn.AssertCloseErr(closeErr)
+	ftpfile.dataconn = fakedconn
+	dc, err = getDataConn(context.Background(), ftpfile, types.OpenRead)
+	s.Error(err, "error is expected")
+	s.ErrorIs(err, closeErr, "error is right kind of error")
+	s.Nil(dc, "dataconn should be nil on error")
+
+	// open dataconn for write after dataconn for read exists
+	ftpfile.dataconn = &dataConn{
+		mode: types.OpenRead,
+		R:    io.NopCloser(strings.NewReader("")),
+	}
+	client.EXPECT().
+		StorFrom(ftpfile.Path(), mock.Anything, uint64(0)).
+		Return(nil).
+		Once()
+	dc, err = getDataConn(context.Background(), ftpfile, types.OpenWrite)
+	s.NoError(err, "no error expected")
+	s.IsTypef(&dataConn{}, dc, "dataconn returned")
+
+	// brief sleep to ensure goroutines running StorFrom can all complete
+	time.Sleep(50 * time.Millisecond)
```

Comment:
Since the mocked 'StorFrom' calls happen inside of a goroutine spawned by 'getDataConn' I added a sleep here to ensure that the mocks are called before 'AssertExpectations'.


I'm open to ideas on a more elegant way to do this.